File: src/go/fab/internal/status/status.go
package status

import (
	"fmt"
	"strings"
	"time"

	"github.com/wvrdz/fab-kit/src/go/fab/internal/config"
	"github.com/wvrdz/fab-kit/src/go/fab/internal/hooks"
	"github.com/wvrdz/fab-kit/src/go/fab/internal/log"
	sf "github.com/wvrdz/fab-kit/src/go/fab/internal/statusfile"
)

// Valid change types.
var ValidChangeTypes = []string{"feat", "fix", "refactor", "docs", "test", "ci", "chore"}

// Valid states.
var ValidStates = []string{"pending", "active", "ready", "done", "failed", "skipped"}

// Allowed states per stage.
var AllowedStates = map[string][]string{
	"intake":    {"active", "ready", "done"},
	"spec":      {"pending", "active", "ready", "done", "skipped"},
	"tasks":     {"pending", "active", "ready", "done", "skipped"},
	"apply":     {"pending", "active", "ready", "done", "skipped"},
	"review":    {"pending", "active", "ready", "done", "failed", "skipped"},
	"hydrate":   {"pending", "active", "ready", "done", "skipped"},
	"ship":      {"pending", "active", "done", "skipped"},
	"review-pr": {"pending", "active", "done", "failed", "skipped"},
}

// Transition defines a state machine transition.
type Transition struct {
	From []string
	To   string
}

// Default transitions (applicable to all stages unless overridden).
var defaultTransitions = map[string]Transition{
	"start":   {From: []string{"pending"}, To: "active"},
	"advance": {From: []string{"active"}, To: "ready"},
	"finish":  {From: []string{"active", "ready"}, To: "done"},
	"reset":   {From: []string{"done", "ready", "skipped"}, To: "active"},
	"skip":    {From: []string{"pending", "active"}, To: "skipped"},
}

// Stage-specific overrides (review and review-pr add fail and start-from-failed).
var stageTransitions = map[string]map[string]Transition{
	"review": {
		"start":   {From: []string{"pending", "failed"}, To: "active"},
		"advance": {From: []string{"active"}, To: "ready"},
		"finish":  {From: []string{"active", "ready"}, To: "done"},
		"reset":   {From: []string{"done", "ready", "skipped"}, To: "active"},
		"fail":    {From: []string{"active"}, To: "failed"},
	},
	"review-pr": {
		"start":   {From: []string{"pending", "failed"}, To: "active"},
		"advance": {From: []string{"active"}, To: "ready"},
		"finish":  {From: []string{"active", "ready"}, To: "done"},
		"reset":   {From: []string{"done", "ready", "skipped"}, To: "active"},
		"fail":    {From: []string{"active"}, To: "failed"},
	},
}

// lookupTransition finds the target state for an event on a stage with a given current state.
func lookupTransition(event, stage, currentState string) (string, error) {
	// Check stage-specific overrides first
	if stageT, ok := stageTransitions[stage]; ok {
		if t, ok := stageT[event]; ok {
			if contains(t.From, currentState) {
				return t.To, nil
			}
			return "", fmt.Errorf("Cannot %s stage '%s' — current state is '%s', no valid transition", event, stage, currentState)
		}
	}

	// Check default transitions
	if t, ok := defaultTransitions[event]; ok {
		if contains(t.From, currentState) {
			return t.To, nil
		}
	}

	return "", fmt.Errorf("Cannot %s stage '%s' — current state is '%s', no valid transition", event, stage, currentState)
}

// Start transitions a stage from {pending,failed} to active.
// If a pre hook is configured for the stage, it runs before the transition.
// A failing pre hook blocks the stage from starting.
func Start(statusFile *sf.StatusFile, statusPath, fabRoot, stage, driver, from, reason string) error {
	if !isValidStage(stage) {
		return fmt.Errorf("Invalid stage '%s'", stage)
	}

	currentState := statusFile.GetProgress(stage)
	targetState, err := lookupTransition("start", stage, currentState)
	if err != nil {
		return err
	}

	// Run pre hook before transitioning
	if err := runStageHook(fabRoot, stage, "pre"); err != nil {
		return err
	}

	statusFile.SetProgress(stage, targetState)
	applyMetricsSideEffect(statusFile, fabRoot, stage, targetState, driver, from, reason)

	return statusFile.Save(statusPath)
}

// Advance transitions a stage from active to ready.
func Advance(statusFile *sf.StatusFile, statusPath, stage, driver string) error {
	if !isValidStage(stage) {
		return fmt.Errorf("Invalid stage '%s'", stage)
	}

	currentState := statusFile.GetProgress(stage)
	targetState, err := lookupTransition("advance", stage, currentState)
	if err != nil {
		return err
	}

	statusFile.SetProgress(stage, targetState)

	return statusFile.Save(statusPath)
}

// Finish transitions a stage to done and auto-activates the next pending stage.
// If a post hook is configured for the stage, it runs after the transition.
// A failing post hook causes the stage to fail.
func Finish(statusFile *sf.StatusFile, statusPath, fabRoot, stage, driver string) error {
	if !isValidStage(stage) {
		return fmt.Errorf("Invalid stage '%s'", stage)
	}

	currentState := statusFile.GetProgress(stage)
	targetState, err := lookupTransition("finish", stage, currentState)
	if err != nil {
		return err
	}

	statusFile.SetProgress(stage, targetState)
	applyMetricsSideEffect(statusFile, fabRoot, stage, targetState, "", "", "")

	// Auto-activate next pending stage
	nextStage := sf.NextStage(stage)
	if nextStage != "" {
		nextState := statusFile.GetProgress(nextStage)
		if nextState == "pending" {
			statusFile.SetProgress(nextStage, "active")
			applyMetricsSideEffect(statusFile, fabRoot, nextStage, "active", driver, "", "")
		}
	}

	if err := statusFile.Save(statusPath); err != nil {
		return err
	}

	// Run post hook after transition is saved
	if err := runStageHook(fabRoot, stage, "post"); err != nil {
		return err
	}

	// Auto-log review/review-pr pass
	if stage == "review" || stage == "review-pr" {
		_ = log.Review(fabRoot, statusFile.Name, "passed", "")
	}

	return nil
}

// Reset transitions a stage to active and cascades downstream to pending.
func Reset(statusFile *sf.StatusFile, statusPath, fabRoot, stage, driver, from, reason string) error {
	if !isValidStage(stage) {
		return fmt.Errorf("Invalid stage '%s'", stage)
	}

	currentState := statusFile.GetProgress(stage)
	targetState, err := lookupTransition("reset", stage, currentState)
	if err != nil {
		return err
	}

	statusFile.SetProgress(stage, targetState)
	applyMetricsSideEffect(statusFile, fabRoot, stage, targetState, driver, from, reason)

	// Cascade downstream to pending
	foundTarget := false
	for _, s := range sf.StageOrder {
		if foundTarget {
			statusFile.SetProgress(s, "pending")
			applyMetricsSideEffect(statusFile, fabRoot, s, "pending", "", "", "")
		}
		if s == stage {
			foundTarget = true
		}
	}

	return statusFile.Save(statusPath)
}

// Skip transitions a stage to skipped and cascades downstream pending to skipped.
func Skip(statusFile *sf.StatusFile, statusPath, fabRoot, stage, driver string) error {
	if !isValidStage(stage) {
		return fmt.Errorf("Invalid stage '%s'", stage)
	}

	currentState := statusFile.GetProgress(stage)
	targetState, err := lookupTransition("skip", stage, currentState)
	if err != nil {
		return err
	}

	statusFile.SetProgress(stage, targetState)
	applyMetricsSideEffect(statusFile, fabRoot, stage, targetState, "", "", "")

	// Forward cascade: downstream pending → skipped
	foundTarget := false
	for _, s := range sf.StageOrder {
		if foundTarget {
			if statusFile.GetProgress(s) == "pending" {
				statusFile.SetProgress(s, "skipped")
				applyMetricsSideEffect(statusFile, fabRoot, s, "skipped", "", "", "")
			}
		}
		if s == stage {
			foundTarget = true
		}
	}

	return statusFile.Save(statusPath)
}

// Fail transitions a stage to failed (review/review-pr only).
func Fail(statusFile *sf.StatusFile, statusPath, fabRoot, stage, driver, rework string) error {
	if !isValidStage(stage) {
		return fmt.Errorf("Invalid stage '%s'", stage)
	}

	currentState := statusFile.GetProgress(stage)
	targetState, err := lookupTransition("fail", stage, currentState)
	if err != nil {
		return err
	}

	statusFile.SetProgress(stage, targetState)

	if err := statusFile.Save(statusPath); err != nil {
		return err
	}

	// Auto-log review/review-pr failure
	if stage == "review" || stage == "review-pr" {
		_ = log.Review(fabRoot, statusFile.Name, "failed", rework)
	}

	return nil
}

// SetChangeType sets the change_type field.
func SetChangeType(statusFile *sf.StatusFile, statusPath, changeType string) error {
	valid := false
	for _, t := range ValidChangeTypes {
		if t == changeType {
			valid = true
			break
		}
	}
	if !valid {
		return fmt.Errorf("Invalid change type '%s' (valid: %s)", changeType, strings.Join(ValidChangeTypes, ", "))
	}
	statusFile.ChangeType = changeType
	return statusFile.Save(statusPath)
}

// SetChecklist updates a checklist field.
func SetChecklist(statusFile *sf.StatusFile, statusPath, field, value string) error {
	switch field {
	case "generated":
		if value != "true" && value != "false" {
			return fmt.Errorf("Invalid value '%s' for field 'generated' (expected true/false)", value)
		}
		statusFile.Checklist.Generated = value == "true"
	case "completed":
		n, err := parseInt(value)
		if err != nil {
			return fmt.Errorf("Invalid value '%s' for field 'completed' (expected non-negative integer)", value)
		}
		statusFile.Checklist.Completed = n
	case "total":
		n, err := parseInt(value)
		if err != nil {
			return fmt.Errorf("Invalid value '%s' for field 'total' (expected non-negative integer)", value)
		}
		statusFile.Checklist.Total = n
	default:
		return fmt.Errorf("Invalid checklist field '%s' (expected: generated, completed, total)", field)
	}
	return statusFile.Save(statusPath)
}

// SetConfidence replaces the confidence block.
func SetConfidence(statusFile *sf.StatusFile, statusPath string, certain, confident, tentative, unresolved int, score float64, indicative bool) error {
	statusFile.Confidence = sf.Confidence{
		Certain:    certain,
		Confident:  confident,
		Tentative:  tentative,
		Unresolved: unresolved,
		Score:      score,
	}
	if indicative {
		statusFile.Confidence.Indicative = sf.BoolPtr(true)
	}
	return statusFile.Save(statusPath)
}

// SetConfidenceFuzzy replaces the confidence block with dimension data.
func SetConfidenceFuzzy(statusFile *sf.StatusFile, statusPath string, certain, confident, tentative, unresolved int, score, meanS, meanR, meanA, meanD float64, indicative bool) error {
	statusFile.Confidence = sf.Confidence{
		Certain:    certain,
		Confident:  confident,
		Tentative:  tentative,
		Unresolved: unresolved,
		Score:      score,
		Fuzzy:      sf.BoolPtr(true),
		Dimensions: &sf.Dimensions{
			Signal:         meanS,
			Reversibility:  meanR,
			Competence:     meanA,
			Disambiguation: meanD,
		},
	}
	if indicative {
		statusFile.Confidence.Indicative = sf.BoolPtr(true)
	}
	return statusFile.Save(statusPath)
}

// AddIssue appends an issue ID (idempotent).
func AddIssue(statusFile *sf.StatusFile, statusPath, id string) error {
	for _, existing := range statusFile.Issues {
		if existing == id {
			return statusFile.Save(statusPath) // refresh last_updated
		}
	}
	statusFile.Issues = append(statusFile.Issues, id)
	return statusFile.Save(statusPath)
}

// AddPR appends a PR URL (idempotent).
func AddPR(statusFile *sf.StatusFile, statusPath, url string) error {
	for _, existing := range statusFile.PRs {
		if existing == url {
			return statusFile.Save(statusPath)
		}
	}
	statusFile.PRs = append(statusFile.PRs, url)
	return statusFile.Save(statusPath)
}

// ProgressMap returns stage:state pairs in pipeline order.
func ProgressMap(statusFile *sf.StatusFile) []sf.StageState {
	return statusFile.GetProgressMap()
}

// ProgressLine returns a single-line visual progress string.
func ProgressLine(statusFile *sf.StatusFile) string {
	var parts []string
	hasActive := false
	hasPending := false

	for _, ss := range statusFile.GetProgressMap() {
		switch ss.State {
		case "done":
			parts = append(parts, ss.Stage)
		case "active":
			parts = append(parts, ss.Stage+" ⏳")
			hasActive = true
		case "ready":
			parts = append(parts, ss.Stage+" ◷")
		case "failed":
			parts = append(parts, ss.Stage+" ✗")
		case "skipped":
			parts = append(parts, ss.Stage+" ⏭")
		case "pending":
			hasPending = true
		}
	}

	if len(parts) == 0 {
		return ""
	}

	line := strings.Join(parts, " → ")
	if !hasActive && !hasPending {
		line += " ✓"
	}
	return line
}

// CurrentStage determines the active/next stage.
func CurrentStage(statusFile *sf.StatusFile) string {
	pm := statusFile.GetProgressMap()

	// First active or ready
	for _, ss := range pm {
		if ss.State == "active" || ss.State == "ready" {
			return ss.Stage
		}
	}

	// Fallback: first pending after last done/skipped
	lastDone := ""
	for _, ss := range pm {
		if ss.State == "done" || ss.State == "skipped" {
			lastDone = ss.Stage
		}
	}

	if lastDone != "" {
		foundLast := false
		for _, ss := range pm {
			if foundLast && ss.State == "pending" {
				return ss.Stage
			}
			if ss.Stage == lastDone {
				foundLast = true
			}
		}
	}

	return "review-pr" // all done
}

// DisplayStage returns the display stage and state as "stage:state".
func DisplayStage(statusFile *sf.StatusFile) (string, string) {
	pm := statusFile.GetProgressMap()

	// Tier 1: first active
	for _, ss := range pm {
		if ss.State == "active" {
			return ss.Stage, "active"
		}
	}

	// Tier 2: first ready
	for _, ss := range pm {
		if ss.State == "ready" {
			return ss.Stage, "ready"
		}
	}

	// Tier 3: last done/skipped
	lastDone := ""
	lastDoneState := ""
	for _, ss := range pm {
		if ss.State == "done" || ss.State == "skipped" {
			lastDone = ss.Stage
			lastDoneState = ss.State
		}
	}
	if lastDone != "" {
		return lastDone, lastDoneState
	}

	// Tier 4: first pending
	if len(sf.StageOrder) > 0 {
		return sf.StageOrder[0], "pending"
	}
	return "intake", "pending"
}

// Validate validates a .status.yaml against the schema.
func Validate(statusFile *sf.StatusFile) error {
	activeCount := 0
	var errors []string

	for _, stage := range sf.StageOrder {
		state := statusFile.GetProgress(stage)
		if state == "" {
			state = "pending"
		}

		if !isValidState(state) {
			errors = append(errors, fmt.Sprintf("Invalid state '%s' for stage %s", state, stage))
			continue
		}

		allowed, ok := AllowedStates[stage]
		if ok && !contains(allowed, state) {
			errors = append(errors, fmt.Sprintf("State '%s' not allowed for stage %s", state, stage))
		}

		if state == "active" {
			activeCount++
		}
	}

	if activeCount > 1 {
		errors = append(errors, "Multiple stages are active (expected 0 or 1)")
	}

	if len(errors) > 0 {
		return fmt.Errorf("%s", strings.Join(errors, "; "))
	}
	return nil
}

// AllStages returns all stage IDs in pipeline order.
func AllStages() []string {
	return sf.StageOrder
}

func applyMetricsSideEffect(statusFile *sf.StatusFile, fabRoot, stage, state, driver, from, reason string) {
	now := time.Now().UTC().Format(time.RFC3339)

	switch state {
	case "active":
		sm, ok := statusFile.StageMetrics[stage]
		if !ok {
			sm = &sf.StageMetric{}
			statusFile.StageMetrics[stage] = sm
		}
		sm.Iterations++
		sm.StartedAt = now
		sm.Driver = driver
		sm.CompletedAt = ""

		// Log transition (best-effort)
		folder := statusFile.Name
		action := "enter"
		if sm.Iterations > 1 {
			action = "re-entry"
		}
		_ = log.Transition(fabRoot, folder, stage, action, from, reason, driver)

	case "done":
		sm, ok := statusFile.StageMetrics[stage]
		if ok {
			sm.CompletedAt = now
		}

	case "pending", "skipped":
		delete(statusFile.StageMetrics, stage)
	}
}

func isValidStage(stage string) bool {
	for _, s := range sf.StageOrder {
		if s == stage {
			return true
		}
	}
	return false
}

func isValidState(state string) bool {
	for _, s := range ValidStates {
		if s == state {
			return true
		}
	}
	return false
}

func contains(slice []string, item string) bool {
	for _, s := range slice {
		if s == item {
			return true
		}
	}
	return false
}

// runStageHook loads the project config and runs the pre or post hook for the given stage.
// Returns nil if no hook is configured or if the hook succeeds.
func runStageHook(fabRoot, stage, phase string) error {
	cfg, err := config.Load(fabRoot)
	if err != nil {
		return fmt.Errorf("failed to load config for stage hooks: %w", err)
	}

	hook := cfg.GetStageHook(stage)
	var command string
	switch phase {
	case "pre":
		command = hook.Pre
	case "post":
		command = hook.Post
	}

	return hooks.Run(fabRoot, command)
}

func parseInt(s string) (int, error) {
	if len(s) == 0 {
		return 0, fmt.Errorf("empty string")
	}
	n := 0
	for _, c := range s {
		if c < '0' || c > '9' {
			return 0, fmt.Errorf("not a number")
		}
		n = n*10 + int(c-'0')
	}
	return n, nil
}

File: src/rust/fab/src/status.rs
use anyhow::{bail, Result};

use crate::config;
use crate::hooks;
use crate::log;
use crate::statusfile::{Confidence, Dimensions, StageMetric, StageState, StatusFile};
use crate::types::{self, STAGE_ORDER, VALID_CHANGE_TYPES};

struct Transition {
    from: Vec<&'static str>,
    to: &'static str,
}

fn default_transitions(event: &str) -> Option<Transition> {
    match event {
        "start" => Some(Transition { from: vec!["pending"], to: "active" }),
        "advance" => Some(Transition { from: vec!["active"], to: "ready" }),
        "finish" => Some(Transition { from: vec!["active", "ready"], to: "done" }),
        "reset" => Some(Transition { from: vec!["done", "ready", "skipped"], to: "active" }),
        "skip" => Some(Transition { from: vec!["pending", "active"], to: "skipped" }),
        _ => None,
    }
}

fn stage_transitions(stage: &str, event: &str) -> Option<Transition> {
    match (stage, event) {
        ("review", "start") | ("review-pr", "start") => {
            Some(Transition { from: vec!["pending", "failed"], to: "active" })
        }
        ("review", "advance") | ("review-pr", "advance") => {
            Some(Transition { from: vec!["active"], to: "ready" })
        }
        ("review", "finish") | ("review-pr", "finish") => {
            Some(Transition { from: vec!["active", "ready"], to: "done" })
        }
        ("review", "reset") | ("review-pr", "reset") => {
            Some(Transition { from: vec!["done", "ready", "skipped"], to: "active" })
        }
        ("review", "fail") | ("review-pr", "fail") => {
            Some(Transition { from: vec!["active"], to: "failed" })
        }
        _ => None,
    }
}

fn lookup_transition(event: &str, stage: &str, current_state: &str) -> Result<String> {
    // Check stage-specific overrides first
    if let Some(t) = stage_transitions(stage, event) {
        if t.from.contains(&current_state) {
            return Ok(t.to.to_string());
        }
        bail!(
            "Cannot {} stage '{}' — current state is '{}', no valid transition",
            event, stage, current_state
        );
    }

    // Check default transitions
    if let Some(t) = default_transitions(event) {
        if t.from.contains(&current_state) {
            return Ok(t.to.to_string());
        }
    }

    bail!(
        "Cannot {} stage '{}' — current state is '{}', no valid transition",
        event, stage, current_state
    );
}

/// Start transitions a stage from {pending,failed} to active.
pub fn start(
    sf: &mut StatusFile,
    status_path: &str,
    fab_root: &str,
    stage: &str,
    driver: &str,
    from: &str,
    reason: &str,
) -> Result<()> {
    if !types::is_valid_stage(stage) {
        bail!("Invalid stage '{}'", stage);
    }

    let current_state = sf.get_progress(stage);
    let target_state = lookup_transition("start", stage, &current_state)?;

    // Run pre hook before transitioning
    run_stage_hook(fab_root, stage, "pre")?;

    sf.set_progress(stage, &target_state);
    apply_metrics_side_effect(sf, fab_root, stage, &target_state, driver, from, reason);

    sf.save(status_path)
}

/// Advance transitions a stage from active to ready.
pub fn advance(
    sf: &mut StatusFile,
    status_path: &str,
    stage: &str,
    _driver: &str,
) -> Result<()> {
    if !types::is_valid_stage(stage) {
        bail!("Invalid stage '{}'", stage);
    }

    let current_state = sf.get_progress(stage);
    let target_state = lookup_transition("advance", stage, &current_state)?;

    sf.set_progress(stage, &target_state);

    sf.save(status_path)
}

/// Finish transitions a stage to done and auto-activates the next pending stage.
pub fn finish(
    sf: &mut StatusFile,
    status_path: &str,
    fab_root: &str,
    stage: &str,
    driver: &str,
) -> Result<()> {
    if !types::is_valid_stage(stage) {
        bail!("Invalid stage '{}'", stage);
    }

    let current_state = sf.get_progress(stage);
    let target_state = lookup_transition("finish", stage, &current_state)?;

    sf.set_progress(stage, &target_state);
    apply_metrics_side_effect(sf, fab_root, stage, &target_state, "", "", "");

    // Auto-activate next pending stage
    if let Some(next_stage) = types::next_stage(stage) {
        let next_state = sf.get_progress(next_stage);
        if next_state == "pending" {
            sf.set_progress(next_stage, "active");
            apply_metrics_side_effect(sf, fab_root, next_stage, "active", driver, "", "");
        }
    }

    sf.save(status_path)?;

    // Run post hook after transition is saved
    run_stage_hook(fab_root, stage, "post")?;

    // Auto-log review/review-pr pass
    if stage == "review" || stage == "review-pr" {
        let _ = log::review(fab_root, &sf.name, "passed", "");
    }

    Ok(())
}

/// Reset transitions a stage to active and cascades downstream to pending.
pub fn reset(
    sf: &mut StatusFile,
    status_path: &str,
    fab_root: &str,
    stage: &str,
    driver: &str,
    from: &str,
    reason: &str,
) -> Result<()> {
    if !types::is_valid_stage(stage) {
        bail!("Invalid stage '{}'", stage);
    }

    let current_state = sf.get_progress(stage);
    let target_state = lookup_transition("reset", stage, &current_state)?;

    sf.set_progress(stage, &target_state);
    apply_metrics_side_effect(sf, fab_root, stage, &target_state, driver, from, reason);

    // Cascade downstream to pending
    let mut found_target = false;
    for s in STAGE_ORDER {
        if found_target {
            sf.set_progress(s, "pending");
            apply_metrics_side_effect(sf, fab_root, s, "pending", "", "", "");
        }
        if *s == stage {
            found_target = true;
        }
    }

    sf.save(status_path)
}

/// Skip transitions a stage to skipped and cascades downstream pending to skipped.
pub fn skip(
    sf: &mut StatusFile,
    status_path: &str,
    fab_root: &str,
    stage: &str,
    _driver: &str,
) -> Result<()> {
    if !types::is_valid_stage(stage) {
        bail!("Invalid stage '{}'", stage);
    }

    let current_state = sf.get_progress(stage);
    let target_state = lookup_transition("skip", stage, &current_state)?;

    sf.set_progress(stage, &target_state);
    apply_metrics_side_effect(sf, fab_root, stage, &target_state, "", "", "");

    // Forward cascade: downstream pending → skipped
    let mut found_target = false;
    for s in STAGE_ORDER {
        if found_target {
            if sf.get_progress(s) == "pending" {
                sf.set_progress(s, "skipped");
                apply_metrics_side_effect(sf, fab_root, s, "skipped", "", "", "");
            }
        }
        if *s == stage {
            found_target = true;
        }
    }

    sf.save(status_path)
}

/// Fail transitions a stage to failed (review/review-pr only).
pub fn fail(
    sf: &mut StatusFile,
    status_path: &str,
    fab_root: &str,
    stage: &str,
    _driver: &str,
    rework: &str,
) -> Result<()> {
    if !types::is_valid_stage(stage) {
        bail!("Invalid stage '{}'", stage);
    }

    let current_state = sf.get_progress(stage);
    let target_state = lookup_transition("fail", stage, &current_state)?;

    sf.set_progress(stage, &target_state);

    sf.save(status_path)?;

    // Auto-log review/review-pr failure
    if stage == "review" || stage == "review-pr" {
        let _ = log::review(fab_root, &sf.name, "failed", rework);
    }

    Ok(())
}

/// Sets the change_type field.
pub fn set_change_type(sf: &mut StatusFile, status_path: &str, change_type: &str) -> Result<()> {
    if !VALID_CHANGE_TYPES.contains(&change_type) {
        bail!(
            "Invalid change type '{}' (valid: {})",
            change_type,
            VALID_CHANGE_TYPES.join(", ")
        );
    }
    sf.change_type = change_type.to_string();
    sf.save(status_path)
}

/// Updates a checklist field.
pub fn set_checklist(
    sf: &mut StatusFile,
    status_path: &str,
    field: &str,
    value: &str,
) -> Result<()> {
    match field {
        "generated" => {
            if value != "true" && value != "false" {
                bail!(
                    "Invalid value '{}' for field 'generated' (expected true/false)",
                    value
                );
            }
            sf.checklist.generated = value == "true";
        }
        "completed" => {
            let n: i64 = parse_non_negative_int(value).map_err(|_| {
                anyhow::anyhow!(
                    "Invalid value '{}' for field 'completed' (expected non-negative integer)",
                    value
                )
            })?;
            sf.checklist.completed = n;
        }
        "total" => {
            let n: i64 = parse_non_negative_int(value).map_err(|_| {
                anyhow::anyhow!(
                    "Invalid value '{}' for field 'total' (expected non-negative integer)",
                    value
                )
            })?;
            sf.checklist.total = n;
        }
        _ => {
            bail!(
                "Invalid checklist field '{}' (expected: generated, completed, total)",
                field
            );
        }
    }
    sf.save(status_path)
}

/// Replaces the confidence block.
pub fn set_confidence(
    sf: &mut StatusFile,
    status_path: &str,
    certain: i64,
    confident: i64,
    tentative: i64,
    unresolved: i64,
    score: f64,
    indicative: bool,
) -> Result<()> {
    sf.confidence = Confidence {
        certain,
        confident,
        tentative,
        unresolved,
        score,
        indicative: if indicative { Some(true) } else { None },
        fuzzy: None,
        dimensions: None,
    };
    sf.save(status_path)
}

/// Replaces the confidence block with dimension data.
pub fn set_confidence_fuzzy(
    sf: &mut StatusFile,
    status_path: &str,
    certain: i64,
    confident: i64,
    tentative: i64,
    unresolved: i64,
    score: f64,
    mean_s: f64,
    mean_r: f64,
    mean_a: f64,
    mean_d: f64,
    indicative: bool,
) -> Result<()> {
    sf.confidence = Confidence {
        certain,
        confident,
        tentative,
        unresolved,
        score,
        indicative: if indicative { Some(true) } else { None },
        fuzzy: Some(true),
        dimensions: Some(Dimensions {
            signal: mean_s,
            reversibility: mean_r,
            competence: mean_a,
            disambiguation: mean_d,
        }),
    };
    sf.save(status_path)
}

/// Appends an issue ID (idempotent).
pub fn add_issue(sf: &mut StatusFile, status_path: &str, id: &str) -> Result<()> {
    if !sf.issues.contains(&id.to_string()) {
        sf.issues.push(id.to_string());
    }
    sf.save(status_path) // always refresh last_updated
}

/// Appends a PR URL (idempotent).
pub fn add_pr(sf: &mut StatusFile, status_path: &str, url: &str) -> Result<()> {
    if !sf.prs.contains(&url.to_string()) {
        sf.prs.push(url.to_string());
    }
    sf.save(status_path) // always refresh last_updated
}

/// Returns stage:state pairs in pipeline order.
pub fn progress_map(sf: &StatusFile) -> Vec<StageState> {
    sf.get_progress_map()
}

/// Returns a single-line visual progress string.
pub fn progress_line(sf: &StatusFile) -> String {
    let mut parts = Vec::new();
    let mut has_active = false;
    let mut has_pending = false;

    for ss in sf.get_progress_map() {
        match ss.state.as_str() {
            "done" => parts.push(ss.stage.clone()),
            "active" => {
                parts.push(format!("{} \u{23f3}", ss.stage));
                has_active = true;
            }
            "ready" => parts.push(format!("{} \u{25f7}", ss.stage)),
            "failed" => parts.push(format!("{} \u{2717}", ss.stage)),
            "skipped" => parts.push(format!("{} \u{23ed}", ss.stage)),
            "pending" => {
                has_pending = true;
            }
            _ => {}
        }
    }

    if parts.is_empty() {
        return String::new();
    }

    let mut line = parts.join(" \u{2192} ");
    if !has_active && !has_pending {
        line.push_str(" \u{2713}");
    }
    line
}

/// Determines the active/next stage.
pub fn current_stage(sf: &StatusFile) -> String {
    let pm = sf.get_progress_map();

    // First active or ready
    for ss in &pm {
        if ss.state == "active" || ss.state == "ready" {
            return ss.stage.clone();
        }
    }

    // Fallback: first pending after last done/skipped
    let mut last_done = String::new();
    for ss in &pm {
        if ss.state == "done" || ss.state == "skipped" {
            last_done = ss.stage.clone();
        }
    }

    if !last_done.is_empty() {
        let mut found_last = false;
        for ss in &pm {
            if found_last && ss.state == "pending" {
                return ss.stage.clone();
            }
            if ss.stage == last_done {
                found_last = true;
            }
        }
    }

    "review-pr".to_string() // all done
}

/// Returns the display stage and state.
pub fn display_stage(sf: &StatusFile) -> (String, String) {
    let pm = sf.get_progress_map();

    // Tier 1: first active
    for ss in &pm {
        if ss.state == "active" {
            return (ss.stage.clone(), "active".to_string());
        }
    }

    // Tier 2: first ready
    for ss in &pm {
        if ss.state == "ready" {
            return (ss.stage.clone(), "ready".to_string());
        }
    }

    // Tier 3: last done/skipped
    let mut last_done = String::new();
    let mut last_done_state = String::new();
    for ss in &pm {
        if ss.state == "done" || ss.state == "skipped" {
            last_done = ss.stage.clone();
            last_done_state = ss.state.clone();
        }
    }
    if !last_done.is_empty() {
        return (last_done, last_done_state);
    }

    // Tier 4: first pending
    if !STAGE_ORDER.is_empty() {
        return (STAGE_ORDER[0].to_string(), "pending".to_string());
    }
    ("intake".to_string(), "pending".to_string())
}

/// Validates a .status.yaml against the schema.
pub fn validate(sf: &StatusFile) -> Result<()> {
    let mut active_count = 0;
    let mut errors = Vec::new();

    for stage in STAGE_ORDER {
        let state = sf.get_progress(stage);
        let state = if state.is_empty() { "pending".to_string() } else { state };

        if !types::is_valid_state(&state) {
            errors.push(format!("Invalid state '{}' for stage {}", state, stage));
            continue;
        }

        if let Some(allowed) = types::allowed_states(stage) {
            if !allowed.contains(&state.as_str()) {
                errors.push(format!("State '{}' not allowed for stage {}", state, stage));
            }
        }

        if state == "active" {
            active_count += 1;
        }
    }

    if active_count > 1 {
        errors.push("Multiple stages are active (expected 0 or 1)".to_string());
    }

    if !errors.is_empty() {
        bail!("{}", errors.join("; "));
    }
    Ok(())
}

/// Returns all stage IDs in pipeline order.
pub fn all_stages() -> &'static [&'static str] {
    STAGE_ORDER
}

fn apply_metrics_side_effect(
    sf: &mut StatusFile,
    fab_root: &str,
    stage: &str,
    state: &str,
    driver: &str,
    from: &str,
    reason: &str,
) {
    let now = chrono::Utc::now().to_rfc3339();

    match state {
        "active" => {
            let sm = sf
                .stage_metrics
                .entry(stage.to_string())
                .or_insert_with(|| StageMetric {
                    started_at: String::new(),
                    driver: String::new(),
                    iterations: 0,
                    completed_at: String::new(),
                });
            sm.iterations += 1;
            sm.started_at = now;
            sm.driver = driver.to_string();
            sm.completed_at = String::new();

            // Log transition (best-effort)
            let folder = &sf.name;
            let action = if sm.iterations > 1 { "re-entry" } else { "enter" };
            let _ = log::transition(fab_root, folder, stage, action, from, reason, driver);
        }
        "done" => {
            if let Some(sm) = sf.stage_metrics.get_mut(stage) {
                sm.completed_at = now;
            }
        }
        "pending" | "skipped" => {
            sf.stage_metrics.remove(stage);
        }
        _ => {}
    }
}

fn run_stage_hook(fab_root: &str, stage: &str, phase: &str) -> Result<()> {
    let cfg = config::load(fab_root)?;
    let hook = cfg.get_stage_hook(stage);
    let command = match phase {
        "pre" => &hook.pre,
        "post" => &hook.post,
        _ => return Ok(()),
    };
    hooks::run(fab_root, command)
}

fn parse_non_negative_int(s: &str) -> Result<i64, ()> {
    if s.is_empty() {
        return Err(());
    }
    for c in s.chars() {
        if !c.is_ascii_digit() {
            return Err(());
        }
    }
    s.parse::<i64>().map_err(|_| ())
}
